src/trace/trace.c

```diff
@@ -96,18 +98,127 @@ static inline void mtrace_event(const char *data, uint32_t length)
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+struct recent_log_entry {
+	uint32_t entry_id;
+	uint64_t message_ts;
+#if CONFIG_TRACE_RECENT_PROLONG
+	uint64_t first_suppression_ts;
+#endif /* CONFIG_TRACE_RECENT_PROLONG */
+	uint32_t suppress_count;
+};
+
+static struct recent_log_entry recent_entries[CONFIG_TRACE_RECENT_ENTRIES_COUNT];
+static struct recent_log_entry entry_to_emit;
+
+static inline void prepare_to_emit(struct recent_log_entry *entry)
+{
+	if (entry->suppress_count)
+		memcpy_s(&entry_to_emit, sizeof(entry_to_emit), entry, sizeof(*entry));
+	memset(entry, 0, sizeof(*entry));
+}
+
+static inline void emit_recent_entry(void)
+{
+	if (!entry_to_emit.suppress_count)
+		return;
+
+	_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
+		     _TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %s",
+		     entry_to_emit.suppress_count, entry_to_emit.entry_id);
+
+	memset(&entry_to_emit, 0, sizeof(entry_to_emit));
+}
+
+static inline void emit_recent_entries(uint64_t current_ts)
+{
+	int i;
+
+	if (entry_to_emit.entry_id)
+		emit_recent_entry();
+
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
+		if (current_ts - recent_entries[i].message_ts >
+			CONFIG_TRACE_RECENT_TIME_THRESHOLD) {
+			prepare_to_emit(&recent_entries[i]);
+			emit_recent_entry();
+		}
+	}
+}
+
+/**
+ * \brief Runtime trace flood suppresion
+ * \param log_level log verbosity level
+ * \param entry ID of log entry
+ * \param message_ts message timestamp
+ * \return false when trace is filtered out, true otherwise
+ */
+static inline bool trace_filter_flood(uint32_t log_level, uint32_t entry, uint64_t message_ts)
+{
+	struct recent_log_entry *oldest_entry = recent_entries;
+	int i;
+
+	/* don't attempt to suppress debug messages using this method, it would be uneffective */
+	if (log_level >= LOG_LEVEL_DEBUG)
+		return true;
+
+	/* check if same log entry was sent recently */
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
+		/* in the meantime identify the oldest entry, for reuse later */
+#if CONFIG_TRACE_RECENT_PROLONG
+		if (oldest_entry->first_suppression_ts > recent_entries[i].first_suppression_ts)
+#else
+		if (oldest_entry->message_ts > recent_entries[i].message_ts)
+#endif /* CONFIG_TRACE_RECENT_PROLONG */
+			oldest_entry = &recent_entries[i];
+
+		if (recent_entries[i].entry_id == entry) {
+#if CONFIG_TRACE_RECENT_PROLONG
+			if (message_ts - recent_entries[i].first_suppression_ts >=
+			    CONFIG_TRACE_RECENT_MAX_TIME) {
+				/* entry was suppressed long enough, time to release it */
+				prepare_to_emit(&recent_entries[i]);
+				return true;
+			} else
+#endif /* CONFIG_TRACE_RECENT_PROLONG */
+			if (message_ts - recent_entries[i].message_ts <
+				   CONFIG_TRACE_RECENT_TIME_THRESHOLD) {
+				recent_entries[i].suppress_count++;
+#if CONFIG_TRACE_RECENT_PROLONG
+				recent_entries[i].message_ts = message_ts;
```

Comment:
looks like this `CONFIG_TRACE_RECENT_PROLONG` is causing a lot of confusion already. Shouldn't `.message_ts` be always set? And `.first_suppression_ts` only if that thing is set? Let's please remove it at least in the first implementation. Getting this right is difficult enough. We can add it later if we really need it. I think always counting time from the most recent message (i.e. "prolonging" in your terminology) is a meaningful thing to do. Let's just do that unconditionally and simplify reviewing and testing.

Reply:
Config flag removed. Hope that will clarify the code a bit.

Reply:
Could you name a specific scenario where that would happen?
Also this would add another config option which you already thought was too complex.

Reply:
the scenario is any quickly repeating trace entry. If a trace entry only repeats twice, or three times in a quick succession, I don't think suppressing already the second message is a good idea. Think about some retrying, or some spurious interrupts or anything. 2-3-4 of them won't kill your system and might be useful for debugging. As for configuration options - yes, I find this PR has too many of them. Not every tunable parameter has to be exported all the way up to the user. Defining several macros in a header should be a much better option IMHO.

Reply:
yeah, I also see the need for a quick burst to be highlighted or at least a message to say they have been suppressed. I think the IRQ is probably a good example (i.e. I'm getting a burst of IRQ traces, so emit one and let the user know this is bursting). This way the user know he's getting a lot of IRQs, but is not flooded by the IRQ traces.
I'm good with tuning in the Kconfig

Reply:
@lgirdwood @lyakh I've implemented it, so that `CONFIG_TRACE_BURST_COUNT` first messages will pass through the filter.
Hope that satisfies your needs.